deckhouse-controller/pkg/registry/deckhouse.go

```diff
@@ -0,0 +1,230 @@
+	"strings"
+	"time"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/ettle/strcase"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	regTransport "github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"gopkg.in/yaml.v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/utils"
+	"github.com/deckhouse/deckhouse/go_lib/dependency"
+	"github.com/deckhouse/deckhouse/go_lib/dependency/cr"
+	"github.com/deckhouse/deckhouse/go_lib/libapi"
+)
+
+type DeckhouseService struct {
+	dc dependency.Container
+
+	registry        string
+	registryOptions []cr.Option
+}
+
+func NewDeckhouseService(registryAddress string, registryConfig *utils.RegistryConfig) *DeckhouseService {
+	return &DeckhouseService{
+		dc:              dependency.NewDependencyContainer(),
+		registry:        registryAddress,
+		registryOptions: utils.GenerateRegistryOptions(registryConfig),
+	}
+}
+
+func (svc *DeckhouseService) GetDeckhouseRelease(ctx context.Context, releaseChannel string) (*ReleaseMetadata, error) {
+	regCli, err := svc.dc.GetRegistryClient(path.Join(svc.registry, "release-channel"), svc.registryOptions...)
+	if err != nil {
+		return nil, fmt.Errorf("get registry client: %w", err)
+	}
+
+	img, err := regCli.Image(ctx, strcase.ToKebab(releaseChannel))
+	if err != nil {
+		if strings.Contains(err.Error(), string(regTransport.ManifestUnknownErrorCode)) {
+			err = errors.Join(err, ErrChannelIsNotFound)
+		}
+
+		return nil, fmt.Errorf("fetch image: %w", err)
+	}
+
+	releaseMetadata, err := svc.fetchReleaseMetadata(img)
+	if err != nil {
+		return nil, fmt.Errorf("fetch release metadata: %w", err)
+	}
+
+	if releaseMetadata.Version == nil {
+		return nil, fmt.Errorf("release metadata malformed: no version found")
+	}
+
+	return releaseMetadata, nil
+}
+
+func (svc *DeckhouseService) ListDeckhouseReleases(ctx context.Context) ([]string, error) {
+	regCli, err := svc.dc.GetRegistryClient(svc.registry, svc.registryOptions...)
+	if err != nil {
+		return nil, fmt.Errorf("get registry client: %w", err)
+	}
+
+	ls, err := regCli.ListTags(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("list tags: %w", err)
+	}
+
+	return ls, nil
+}
+
+type canarySettings struct {
+	Enabled  bool            `json:"enabled"`
+	Waves    uint            `json:"waves"`
+	Interval libapi.Duration `json:"interval"` // in minutes
+}
+
+type ReleaseMetadata struct {
+	Version      *semver.Version           `json:"version"`
+	Canary       map[string]canarySettings `json:"canary"`
+	Requirements map[string]string         `json:"requirements"`
+	Disruptions  map[string][]string       `json:"disruptions"`
+	Suspend      bool                      `json:"suspend"`
+
+	Changelog map[string]interface{}
+
+	Cooldown *metav1.Time `json:"-"`
+}
+
+func (svc *DeckhouseService) fetchReleaseMetadata(image v1.Image) (*ReleaseMetadata, error) {
+	var meta = new(ReleaseMetadata)
+
+	layers, err := image.Layers()
```

Comment:
Ouch, sorry, my bad. It's the wrong way to scan each layer for a desired file. 
Much more better to [fetch flattened filesystem](https://github.com/deckhouse/deckhouse/blob/699630ccd158425168b114fe0f9c94e5cccc2de7/deckhouse-controller/pkg/controller/module-controllers/downloader/module_downloader.go#L361) and [scan for the desired file](https://github.com/deckhouse/deckhouse/blob/699630ccd158425168b114fe0f9c94e5cccc2de7/deckhouse-controller/pkg/controller/module-controllers/downloader/module_downloader.go#L419).

Reply:
i'm sure, it's copy-paste from another place.
keep an eye on it.

Reply:
done